File: java-version/src/main/java/com/example/java_version/service/SampleUserService.java
package com.example.java_version.service;

import com.example.java_version.dto.CreateUserDTO;
import com.example.java_version.dto.UserCreatedDTO;
import com.example.java_version.entity.SampleUser;
import com.example.java_version.exception.EmailAlreadyExistsException;
import com.example.java_version.repository.SampleUserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class SampleUserService {

    private final SampleUserRepository userRepository;

    private final PasswordEncoder passwordEncoder;

    public SampleUserService(SampleUserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    public UserCreatedDTO create(CreateUserDTO dto) {
        if (userRepository.findUserByEmail(dto.email()).isPresent()) {
            throw new EmailAlreadyExistsException("Email já cadastrado na base de dados!");
        }

        String encryptedPassword = passwordEncoder.encode(dto.password());

        SampleUser user = SampleUser.builder()
                .username(dto.username())
                .email(dto.email())
                .password(encryptedPassword)
                .build();

        userRepository.save(user);

        return new UserCreatedDTO("Usuário criado com sucesso!");
    }
}

File: kotlin-version/src/main/kotlin/com/example/kotlin_version/service/SampleUserService.kt
package com.example.kotlin_version.service

import com.example.kotlin_version.dto.CreateUserDTO
import com.example.kotlin_version.dto.UserCreatedDTO
import com.example.kotlin_version.entity.SampleUser
import com.example.kotlin_version.exceptions.EmailAlreadyExistsException
import com.example.kotlin_version.repository.SampleUserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class SampleUserService(
    private val userRepository: SampleUserRepository,
    private val passwordEncoder: PasswordEncoder
){
    fun create(dto: CreateUserDTO): UserCreatedDTO {

        userRepository.findUserByEmail(email = dto.email)?.let {
            throw EmailAlreadyExistsException(message = "Usuário com esse email já existe!")
        }

        val encryptedPassword = passwordEncoder.encode(dto.password)
            ?: throw RuntimeException("Houve um erro ao encriptar senha")

        val user = SampleUser(username = dto.username, password = encryptedPassword, email = dto.email)

        userRepository.save(user)

        return UserCreatedDTO(message = "Usuário criado com sucesso!")
    }

}